### PR TITLE
THREESCALE-1643: Update notification preferences via API

### PR DESCRIPTION
**Pull request by mayorova**
**What this PR does / why we need it**:

Update personal notification preferences via API.

It's still to be decided whether we need to support updating preferences of **another** user (e.g. an admin to be able to change preferences of a member).

**Which issue(s) this PR fixes** 

https://issues.redhat.com/browse/THREESCALE-1643

**Verification steps** 

TBD

**Special notes for your reviewer**:

### CODE REVIEW

File: app/representers/notification_preferences_representer.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class NotificationPreferencesRepresenter < ThreeScale::Representer
+  include ThreeScale::JSONRepresenter
+
+  property :preferences, as: :notification_preferences
```
Comment (mayorova) [inline app/representers/notification_preferences_representer.rb:6]: With this representer the result is shown as:
```
{
    "notification_preferences": {
        "application_created": true,
        "account_created": false,
        "limit_alert_reached_provider": false,
        ...
        "daily_report": false
    }
}
```

I think it's quite in line with most of other APIs that wrap the parameters in a "root" object. But on the other hand, I think it's not necessary, and it might be better to not have such a root object, see below.

File: app/representers/notification_preferences_representer.rb [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+  # NotificationPreferences.available_notifications.each do |notification|
```
Comment (mayorova) [inline app/representers/notification_preferences_representer.rb:8]: With this code, the response will look like:

```
{
    "application_created": true,
    "account_created": false,
    "limit_alert_reached_provider": false,
    ...
    "daily_report": false
}
```

I think it's more similar to the request, where we'd just set form-data `application_created=true&account_created=true` etc.

In fact, I think the output is the same as if we were returning `current_user.notification_preferences.preferences` in the controller. But I think going through a representer is more standard and flexible.

File: app/models/notification_preferences.rb [hunk 3]
```diff
@@ -31,6 +31,11 @@ def preferences=(preferences)
     super Hash(preferences).stringify_keys
   end
 
+  def new_preferences=(updated_preferences)
```
Comment (mayorova) [inline app/models/notification_preferences.rb:34]: The existing `enabled_notifications=` assumes that whatever is NOT provided in the list is disabled. This suits for the UI, where the entry (`notification_preferences[enabled_notifications][]=account_created`) is only added for those checkboxes that are enabled, and indeed others are considered disabled.

For the API, I wanted to give the user the possibility of only toggling individual preferences - if the preference is not provided in the request, it means that it remains the same. That's why I thought that this new method would be good.

It also transforms "true", "false" and other values into `true`, `false` etc.

File: doc/active_docs/account_management_api.json [hunk 4]
```diff
@@ -7251,6 +7251,149 @@
...
+                "type": "object",
+                "properties": {
+                  "access_token": {
+                    "type": "string",
+                    "description": "A personal Access Token."
+                  },
+                  "application_created": {
+                    "type": "boolean",
+                    "description": "Application created."
+                  },
```
Comment (jlledom) [inline doc/active_docs/account_management_api.json:7299]: This kind of definition allows me to click the checkbox "Send empty value":

image

Which sends 'null' instead of `true` or `false`. IMO that's wrong, but the worst part is it actually stores `null` in the DB column:

image

I think `null` shouldn't be allowed, but if we want to allow it, then we should transform it to `false`.

Also, I think we should remove the "Send empty value" box from Swagger UI.

File: app/models/notification_preferences.rb (hunk 3 above)
Comment (jlledom) [reply app/models/notification_preferences.rb:34]: On the other hand, notifications belong to categories, which are enabled or disabled based on some external parameter. When a category is disabled, all notifications in the category are disabled and can't be enabled. UI implements this by not adding to the HTML form any notification belonging to a disabled category, and setting to `false` any notification not received from the form.

This new method allows editing all available notifications, which I think its fine, but it enables received notifications not matter whether it's category is disabled. The categories protection must be enforced in the API endpoint, maybe not particularly in this method but we should do it somewhere.

You can check the categories implementation in the `NotificationPreferencesForm` class.

This is an example: Go to Audience -> Accounts -> Settings -> Usage rules and disable Service plans. Then the next preferences should be always `false`:

```
    "service_contract_created": false,
		"service_plan_change_requested": false,
		"service_contract_plan_changed": false,
		"service_contract_cancellation": false,
```

File: test/integration/admin/api/personal/notification_preferences_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    assert_response :success
+    user.notification_preferences.reload
+    update_params.each_pair do |key, value|
+      assert_equal value, prefs[key.to_s], "Notification '#{key}' should be set to '#{value}'"
+      assert_equal value, user.notification_preferences.include?(key.to_s)
+    end
+  end
+
+  test 'update with wrong parameters' do
```
Comment (jlledom) [inline test/integration/admin/api/personal/notification_preferences_controller_test.rb:44]: IMO we should decide whether allowing `null` or not, and if not, test it here.

File: test/integration/admin/api/personal/notification_preferences_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    get admin_api_personal_notification_preferences_path(format: :json, access_token: @token)
+
+    prefs =  JSON.parse(response.body)['notification_preferences']
+
+    assert_response :success
+    assert_same_elements NotificationPreferences.available_notifications, prefs.keys.map(&:to_sym)
+    assert prefs.values.all? { [true, false].include?(_1) }
+  end
+
+  test 'update successfully' do
```
Comment (jlledom) [inline test/integration/admin/api/personal/notification_preferences_controller_test.rb:25]: I think we should add another test for the categories restriction I explained above.

File: test/models/notification_preferences_test.rb [hunk 7]
```diff
@@ -75,10 +75,28 @@ def test_hidden_preferences
...
 
-    enabled = preferences.enabled_notifications = %w(application_created)
+    enabled = preferences.enabled_notifications = %w[application_created]
     enabled += hidden_notifications.map(&:to_s)
 
     assert_same_elements enabled, preferences.enabled_notifications
     assert_equal preferences.available_notifications, Set.new(preferences.preferences.keys)
   end
+
+  test 'new_preferences=' do
```
Comment (jlledom) [inline test/models/notification_preferences_test.rb:85]: Maybe it should test that all notification preferences not explicitly mentioned are not altered.

File: app/models/notification_preferences.rb (hunk 3 above)
Comment (mayorova) [reply app/models/notification_preferences.rb:34]: Hm, yeah, you're right. I think I wanted to look into the permissions thing, but then forgot :grimacing:  

However, I have a doubt.

> it enables received notifications not matter whether it's category is disabled.

Isn't there a check for permissions on sending notifications?

Because what happens if, e.g:
1. service plans are enabled
2. "service_x" notifications are set to `true`
3. service plans are disabled

Will those notifications be sent?

So, do you think we should hide "service_" notifications from the JSON response? Or ignore the corresponding request params, if service plans are disabled?

File: app/representers/notification_preferences_representer.rb (hunk 1 above)
Comment (mayorova) [reply app/representers/notification_preferences_representer.rb:6]: Examples:

```
curl -X GET 'http://provider-admin.3scale.localhost:3000/admin/api/settings.json?access_token=admintoken'
```

```
{
  "settings": {
    "useraccountarea_enabled": true,
    "signups_enabled": true,
    "account_approval_required": false,
    "strong_passwords_enabled": false,
    "public_search": false,
    "account_plans_ui_visible": true,
    "change_account_plan_permission": "request",
    "service_plans_ui_visible": true,
    "change_service_plan_permission": "request",
    "enforce_sso": false
  }
}
```

```
curl -X GET 'http://provider-admin.3scale.localhost:3000/admin/api/users/55/permissions.json?access_token=admintoken'
```

```
{
  "permissions": {
    "user_id": 55,
    "role": "admin",
    "allowed_sections": [
      "finance",
      "monitoring",
      "partners",
      "plans",
      "policy_registry",
      "portal",
      "settings"
    ],
    "allowed_service_ids": null,
    "links": [
      {
        "rel": "user",
        "href": "http://provider-admin.3scale.localhost/admin/api/accounts/2/users/55"
      }
    ]
  }
}
```

```
curl -X GET 'http://provider-admin.3scale.localhost:3000/admin/api/services/2/features.json?access_token=admintoken'
```

```
{
  "features": [
    {
      "feature": {
        "id": 1,
        "name": "Unlimited Greetings",
        "system_name": "unlimited_greetings",
        "scope": "application_plan",
        "visible": true,
        "created_at": "2024-10-03T15:48:48+05:00",
        "updated_at": "2024-10-03T15:48:48+05:00",
        "links": [
          {
            "rel": "self",
            "href": "http://provider-admin.3scale.localhost/admin/api/services/2/features/1"
          },
          {
            "rel": "service",
            "href": "http://provider-admin.3scale.localhost/admin/api/services.2"
          }
        ]
      }
    }
   ...]
}
```

File: test/integration/admin/api/personal/notification_preferences_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+      assert_equal value, user.notification_preferences.include?(key.to_s)
+    end
+  end
+
+  test 'update with wrong parameters' do
+    user.notification_preferences.update(enabled_notifications: ['account_created'])
+    previous_prefs = user.notification_preferences.preferences
+    update_params = {
+      non_existing_notification: false,
+      account_created: 'some-value'
```
Comment (mayorova) [inline test/integration/admin/api/personal/notification_preferences_controller_test.rb:49]: Actually, this is wrong in the context of this test :upside_down_face: 
As currently I'm using `ActiveModel::Type::Boolean.new.cast`, any value that is not in [this list](https://github.com/rails/rails/blob/v7.0.8.7/activemodel/lib/active_model/type/boolean.rb#L17-L26) is considered `true`*, so `account_created: "some-value"` will effectively set this notification to `true`.

*except empty value that is transformed to `nil`.

Now I'm thinking - maybe it's confusing, and we should actually only limit the values to `false`, `true` and `"false"`, `"true"`??

File: app/models/notification_preferences.rb (hunk 3 above)
Comment (jlledom) [reply app/models/notification_preferences.rb:34]: > Hm, yeah, you're right. I think I wanted to look into the permissions thing, but then forgot 😬
> 
> However, I have a doubt.
> 
> > it enables received notifications not matter whether it's category is disabled.
> 
> Isn't there a check for permissions on sending notifications?
> 
> Because what happens if, e.g:
> 
>     1. service plans are enabled
> 
>     2. "service_x" notifications are set to `true`
> 
>     3. service plans are disabled
> 
> 
> Will those notifications be sent?

I didn't think on this, now I tested this and the answer is notifications are still sent. They are sent until notification preferences are updated, via UI. For consistence, I think updating notifications via API should also have the same effect. So I still think notifications should always be set to false when they belong to a disabled category.

> Or ignore the corresponding request params, if service plans are disabled?

In my opinion the endpoint should always set disabled notifications to false, no matter whether those notifications were received or which value they had. We should only honor values for received notifications if they are enabled. 

> So, do you think we should hide "service_" notifications from the JSON response? 

I don't have a strong opinion, whether we return them or not, I'm fine. So maybe I would return all of them , just to simplify the code.

**Review by jlledom — COMMENTED** [review]
Looks good so far. I'll approve if you:

- [ ] Add the categories filtering
- [ ] Add tests for the categories thing
- [x] Add tests for the `null` value
- [x] The restaurant is good

File: app/models/notification_preferences.rb [hunk 9]
```diff
@@ -31,6 +31,11 @@ def preferences=(preferences)
...
+    prefs = updated_preferences.transform_values { ActiveModel::Type::Boolean.new.cast(_1) || false }
```
[the remark this answers is not in the document]
Comment (mayorova) [reply app/models/notification_preferences.rb]: @jlledom so, I am trying to implement an alternative approach... let me know what you think:

https://github.com/3scale/porta/pull/4080/commits/09ef6ca55fed1ddc732064512aa608f3fcc5547a

File: app/models/notification_preferences.rb (hunk 3 above)
Comment (mayorova) [reply app/models/notification_preferences.rb]: TL;DR; I would keep it as it is :grimacing:

**Review by jlledom — DISMISSED** [review]

File: app/models/notification_preferences.rb [hunk 10]
```diff
@@ -77,4 +88,22 @@ def set_defaults
...
+      true
+    else
+      value
+    end
+  end
+
+  def preferences_valid?
+    preferences.each_pair do |key,value|
+      errors.add(:preferences, :invalid_value, key: key) unless [true, false].include? value
+      errors.add(:preferences, :invalid_key, key: key) unless available_notifications.include? key.to_s
```
[the remark this answers is not in the document]
Comment (mayorova) [reply app/models/notification_preferences.rb:106]: That's right, but the model can potentially be updated through something else, not that controller, so I think the check is still valid.

File: config/locales/en.yml [hunk 11]
```diff
@@ -1301,6 +1301,12 @@ en:
               charging_failed: 'Failed to charge the credit card.'
             period:
               invalid_range: 'must be between the provider account creation date and 12 months from now'
+              
+        notification_preferences:
+          attributes:
+            preferences:
+              invalid_value: "invalid value for '%{key}', must be either true or false"
```
[the remark this answers is not in the document]
Comment (mayorova) [reply config/locales/en.yml:1308]: Same answer :) Not through the API controller, but the model can be updated in a different way (potentially).

File: app/representers/notification_preferences_representer.rb [hunk 12]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class NotificationPreferencesRepresenter < ThreeScale::Representer
+  include ThreeScale::JSONRepresenter
+
+  property :preferences, as: :notification_preferences, getter: ->(*) { preferences.sort.to_h }
```
Comment (mayorova) [inline app/representers/notification_preferences_representer.rb:6]: Sorting in alphabetical order

File: test/models/notification_preferences_test.rb (hunk 7 above)
Comment (mayorova) [reply test/models/notification_preferences_test.rb:85]: I think I addressed it here: https://github.com/3scale/porta/pull/4080/commits/916e3d7151aa04a69a718b7669a4ab4349f03b70

File: app/models/notification_preferences.rb [hunk 13]
```diff
@@ -74,4 +90,22 @@ def set_defaults
...
+  def transform_boolean(value)
+    case value
+    when "false"
+      false
+    when "true"
+      true
+    else
+      value
+    end
+  end
```
Comment (jlledom) [inline app/models/notification_preferences.rb:103]: I'm surprised we need this. Don't we (or Rails) have some util class to handle boolean inputs?

**Review by jlledom — APPROVED** [review]
Works perfect for me, good job! :+1:

File: app/models/notification_preferences.rb (hunk 13 above)
Comment (mayorova) [reply app/models/notification_preferences.rb:103]: Yeah, I was also surprised...

In other occasions I think we use `ActiveModel::Type::Boolean.new.cast` but it acts a bit differently, and (to my liking) a bit confusing.
